src/IO/S3/Credentials.cpp

```diff
@@ -717,8 +722,13 @@ S3CredentialsProviderChain::S3CredentialsProviderChain(
     AddProvider(std::make_shared<Aws::Auth::ProfileConfigFileAWSCredentialsProvider>());
 }
 
+std::string determineAvailabilityZone(const DB::S3::PocoHTTPClientConfiguration & configuration)
+{
+    auto aws_client_configuration = getAwsClientConfig(configuration);
+    auto ec2_metadata_client = InitEC2MetadataClient(aws_client_configuration);
+    return ec2_metadata_client->getCurrentAvailabilityZone();
```

Comment:
I think this should be static variable as it makes sense to initialize it only once. And if it is static, it can also make sense to use `std::variant<std::string, std::exception_ptr>` for it. E.g. something like this:
```
std::variant<std::string, std::exception_ptr> getAvailabilityZoneOrException()
{
    try
     {
            return getAWSEC2MetadataClient()->getCurrentAvailabilityZone();
     }
    catch (...)
    {
        return std::current_exception();
    }
 }

std::string getAvailabilityZone()
{
    static auto az_or_exception = getAvailabilityZoneOrException();
    if (const auto * az = std::get_if<std::string>(&az_or_exception))
        return *az;
    else
        std::rethrow_exception(std::get<std::exception_ptr>(az_or_exception));
}
```

Reply:
Cool. I think it should be a static method as well. Metadata for availability zone should not require credential/token in the HTTP request. Will change to remove the dependency on credential part.

```sh
kubectl exec testing-pod --  curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone
us-west-2b
```

Reply:
I gave it a try, but now we need to iterate AWS, and then GCP, (maybe later Azure). the std::<exception_ptr> does not allow me to get the dispaly_text. I want both exception information can be printed out. Two nested `Try` block does not seem elegant. Let me know if you think of better ways to handle.